### ISSUE TITLE
$dispatch not working (only on last element) when looping over Spruce Arrays

### ISSUE THREAD

**Issue by RichardAnderson** [ryangjchandler/spruce#152]
**Issue Summary:**
When looping using template and x-for an array stored in spruce, where an event needs to be dispatched, and $dispatch is used in-line or via passthrough, events are not dispatched for the last element within the array.

NOTE:  This cannot be replicated when just using Alpine (e.g.: https://jsfiddle.net/racsuk/b94qm023/)

**Replication Steps:**

Full example can be seen in the following JsFiddle:
https://jsfiddle.net/racsuk/9gd6npmc/

Click "Item4" either "Passthrough" or "Inline" (or whatever is the last item in the array)

_Expected Results:_
Console reports "Event called x times"
The "Event has been called x times" is incremented

_Actual Results:_
Event is not run, and counter does not increase

Spruce:
```
window.Spruce.store(
   "data", 
   { 
      items: ["item1", "item2", "item3", "item4"],
      removeItem(index) {
         this.items.splice(index, 1);
      },
   }
);
```

JavaScript:
```
function ParentObject() {
   return {
      called: 0,
    
      save() {
         this.called++;
         console.log("Event called " + this.called + " times");
      },
   }
}

function ChildObject() {
   return {   
      removeItemPassthrough(index, dispatch) {
         this.$store.data.removeItem(index);
         dispatch("save-record");
      },
      removeItem(index) {
         this.$store.data.removeItem(index);
      }
   }
}
```

HTML:
```        
<div x-data="ParentObject()" @save-record.window="save()">
   <div x-data="ChildObject()">
      <template x-for="(item, key) in $store.data.items">
         <div>
            <span x-text="item"></span>
            <a href="#" @click.prevent="removeItemPassthrough(key, $dispatch)">Passthrough</a>
            <a href="#" @click.prevent="removeItem(key); $dispatch('save-record');">Inline</a>
         </div>
      </template>
   </div>
   <div>Event has been called <span x-text="called"></span> times</div>
</div>
```

**Comment by ryangjchandler** [comment]
@RichardAnderson It looks like the problem occurs **only** when `$dispatch` is called **after** removing the item from the array. 

Here's a fork of your fiddle where the order has been swapped, it works fine - https://jsfiddle.net/7h4zc1fy/1/

I'm guessing the reason that this is causing an issue is that when the item is removed the array, Spruce instantaneously tells Alpine to update (hence removing the item from the markup) and then trying to call `$dispatch` will cause some trouble as that element doesn't exist anymore and `$dispatch` will call `elementThatTriggeredEvent.dispatchEvent`.

I'd suggest trying to dispatch **before** removing the item, if that's possible.

**Comment by RichardAnderson** [comment]
@ryangjchandler thanks for the fast response!  Loving Spruce, thanks for all the hard work!

Unfortunately this is not possible - as I am using the data in spruce to submit the items to an API, and dispatching the event before removing the item submits all items to the API.   The reason I feel this is still an issue, is that it works for all elements, consistently I might add, apart form the last. 

One workaround, which kind of negates the advantage of Spruce, which I have used as a workaround in this instance, is on initiation of the Child Control, I copy the item data using JSON.parse(JSON.stringify(this.$store.data.items)) to a local parameter in Alpine, and then bind all my HTML to the new local items array.  This correctly dispatches the events in the right order, but unbinds me from the true source of data in Spruce.

**Comment by ryangjchandler** [comment]
Right, I see now. Let me debug this a little further & see what I can find for you.

**Comment by RichardAnderson** [comment]
Just for reference:
https://user-images.githubusercontent.com/860443/120708922-1925d800-c4b4-11eb-90a1-c95d0ff9e393.mp4

Works for item 1, 2 & 3 but not 4 (or the last element)
